turbinia/lib/recipe_helpers.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+# Copyright 2021 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Library to contain recipe validation logic."""
+
+import copy
+import logging
+import yaml
+from turbinia import config
+from turbinia import TurbiniaException
+from yaml import Loader, load, dump
+from turbinia.lib.file_helpers import file_to_str, file_to_list
+from turbinia.client import TASK_MAP
+
+log = logging.getLogger('turbinia')
+
+#Attributes allowed on the 'globals' task recipe
+DEFAULT_GLOBALS_RECIPE = {
+    'debug_tasks': False,
+    'jobs_allowlist': [],
+    'jobs_denylist': [],
+    'yara_rules': '',
+    'filter_patterns': [],
+    'yara_rules_file': None,
+    'filter_patterns_file': None
+}
+
+#Default 'task_recipes' dict
+DEFAULT_RECIPE = {'globals': DEFAULT_GLOBALS_RECIPE}
+
+
+def load_recipe_from_file(recipe_file):
+  """ Load recipe from file. """
+  if not recipe_file:
+    task_recipe = DEFAULT_RECIPE
+  else:
+    try:
+      with open(recipe_file, 'r') as r_file:
+        recipe_file_contents = r_file.read()
+        recipe_dict = load(recipe_file_contents, Loader=Loader)
+    except yaml.parser.ParserError as exception:
+      message = (
+          'Invalid YAML on recipe file {0:s}: {1!s}.'.format(
+              recipe_file, exception))
+      log.error(message)
+      return False
+    except IOError as exception:
+      log.error(
+          'Failed to read recipe file {0:s}: {1!s}'.format(
+              recipe_file, exception))
+      validate_recipe(recipe_dict)
+      return False
+    return recipe_dict
+
+
+def validate_globals_recipe(proposed_globals_recipe):
+  """Ensures globals recipe is valid for further processing."""
+
+  for item in DEFAULT_GLOBALS_RECIPE:
+    if item not in proposed_globals_recipe:
+      proposed_globals_recipe[item] = DEFAULT_GLOBALS_RECIPE[item]
+
+  filter_patterns_file = proposed_globals_recipe.get(
+      'filter_patterns_file', None)
+  yara_rules_file = proposed_globals_recipe.get('yara_rules_file', None)
```

Comment:
This came up in our previous chat about this PR. There are two values that are part of the globals recipe allowed keys: 'yara_rules' and 'filter_patterns'. The purpose of the function is to validate that nothing major is wrong with the globals task recipe and also ensuring that minor things like an absent field don't get in the way of the submission, as the defaults for the globals are quire reasonable. in the case of those two, it's just empty, but thinking through this it also introduces a dependency on how the file helpers work in case the file is invalid etc.. which brings more opportunity for bugs.

I am changing to the following behaviour:

If no yara / patterns files are given, nothing will be included in the globals task recipe. I am also building a deepcopying a new dict regardless, as I assume we want to keep the logic that auto-completes missing fields with the defaults.

Reply:
Resolving tentatively.